### PR TITLE
WIP: partitioned exact-prefix VDN transport

### PR DESCRIPTION
**Pull request by xmarre**
Development PR for VDN-side partitioned exact-prefix progressive attention transport. Base is released VDN-H3-Plus main.

Current scope includes the fail-closed Flow wire/external-mode contract, heterogeneous grouped VDN softmax execution with provider-v4 mapped query positions, exact-prefix key-measure transport, pre-gather inherited Q/K/V preprocessing, bounded execution-owned gather-index caching, and API-4 `partitioned_attention_variable_grid_linear` execution for the learned linear complement. The variable-grid extension preserves released raw pre-RoPE Q/K/V ownership, `vdn_solve` recurrence, text state, alpha bridge, gates/norm/output projection and anchor-end semantics while running spatial short-conv on each frame's actual grid and mapping only cross-grid temporal taps on MiniMax-H3's own area-normalized spatial coordinate lattice with FP32 bilinear sampling.

Current coordinated validation stack:

```text
Flow #49  3912887372146e0691e6d0b91d99b64710a94767
Sol  #15  93b3e03f2b7b579aaf55fa0f87f55083b259e25c
VDN  #19  675cd6e7b43c9d8d63b7a91834fb6d775388a2ee
```

VDN hosted CI remains green at this exact VDN head. The exact runtime source used by 00508 is preserved as `checkpoint/exact-prefix-sm120-runtime-pass-00508-20260918` across Flow, Sol and VDN.

## 00508 passes the coordinated scheduling/runtime gate

Evidence:

```text
ComfyUI log sha256  dce118a666aa44c2a7cbb7cdcf43aba6ad450ee075e6a73f85a49a26a1038b1b
Flow metrics sha256 c51d6c729a3c08ee85f8546b97a40f74cebeb72277b33c22b77c40179624ee53
```

The continuation is now `9L/7A/2F`: low `5L/4A/1F`, probe `1/1/0`, high `3/2/1`; whole prompt `18L/14A/4F`. Flow reports `2` partition-provider creations and `4` reuses, and the repeated sparse-phase Sol `dense_provider` transitions from the failed runs are gone.

That confirms the previous scheduling failure was not VDN API-4 transport. Flow v6 now uses the same semantic provider identity boundary as Sol history-v1 and rebinds the current preprocessing closure while preserving one visible partitioned provider identity.

## VDN heterogeneous hardware evidence retained

00508 again proves the actual API-4 heterogeneous path:

- exact target-prefix latent is not resized for H3;
- prefix uses target-grid RoPE and generated suffix uses source-grid RoPE;
- suffix is represented by the real source-grid latent;
- grouped VDN softmax and the variable-grid learned linear complement are active;
- mapped rectangular Sol requests execute through the one-union partitioned route;
- requested Q rows exactly equal kernel Q rows;
- square-Q expansion remains zero;
- learned 3D transfer discards its prefix output and restores the authoritative target prefix;
- final protected prefix remains exact and high begins with an actual H3 evaluation.

No VDN source change was required after 00507.

## Evidence-validator correction

00508 exposed a Flow-only offline validator mismatch: provider binding occurs on every logical low/probe model-wrapper entry, including a call later satisfied by Spectrum forecast, while `partitioned_transformer_calls` counts only actual H3 executions. The corrected Flow gate therefore compares provider bindings with logical low/probe calls and transformer executions with actual low/probe calls. This does not change VDN runtime semantics.

## 00529 source-low evidence and physical-grid temporal mapping candidate

The completed matched `00529` run closes the previous source-low discriminator. The upper-45% vertical boundary impulse is already present in `source_low_native` before learned 3D transfer, while `source_low_exact_context` is numerically identical to that receipt to sub-microcell precision. The learned upscaler then amplifies the displacement and exact-prefix restoration increases the immediate seam further. This moves the first demonstrated origin into the partitioned low/probe H3 path rather than the upscaler.

Auditing the variable-grid VDN linear complement found one concrete mixed-grid coordinate mismatch. Its temporal depthwise short-convolution was mapping only the cross-grid taps with:

```python
F.interpolate(..., mode="bilinear", align_corners=False)
```

That is PyTorch's half-pixel image-resize lattice. The same partitioned H3 sequence assigns spatial positions with ComfyUI MiniMax-H3's `_axis_from_sqrt_area` lattice for RoPE. At the actual 00529 patch grids, source `20x27` and target `28x38`, the two mappings are not equivalent: low→target differs by roughly **0.118–0.169 source patch cells vertically** (mean **0.1438**), and target→low differs by roughly **-0.237 to -0.167 target patch cells** (mean **-0.2018**). These are coordinate differences, not a claimed direct prediction of output displacement, but they are the same order as the measured boundary defect and occur only at the heterogeneous temporal boundary.

The current candidate replaces only those cross-grid temporal taps with an explicit `grid_sample` map derived from the exact H3 area-normalized coordinate formula. Same-grid frames still return the original tensor and therefore retain released fixed-grid arithmetic. Cross-grid interpolation remains FP32 and uses border padding for the small endpoint extent mismatch. The resampling grid is cached for the duration of the variable-feature call so K/V feature construction does not rebuild it repeatedly.

Regression coverage now:
- oracles the replicated coordinate formula against pinned ComfyUI `_axis_from_sqrt_area`;
- verifies a linear physical-coordinate field is recovered by the new cross-grid map;
- proves the previous half-pixel resize is measurably different on heterogeneous H3 geometry;
- preserves the exact identity path for equal grids;
- keeps the existing uniform-grid VDN oracle unchanged.

This is a **candidate causal correction**, not yet a validated visual fix. CPU/source CI proves the mapping and unchanged uniform-grid contract; only the matched CUDA/decoded rerun can establish whether it removes the observed framing jump without a quality regression. Development run `35465375133` and exact-head CI run `35465587977` are fully green. Checkpoint: `checkpoint/pr19-h3-physical-temporal-map-pr-head-green-20260919`.

## Remaining gates

VDN's heterogeneous API-4 hardware gate and the coordinated provider-lifetime/Spectrum scheduling gate are passed. Visual continuity is still blocked pending a matched run of the physical-grid temporal-map candidate above. High-stage memory/performance is tracked separately in stacked PR #23. Ordinary grouped/native VDN execution and unrelated VDN PR #8 remain untouched.

Draft remains appropriate until those final acceptance gates pass.

## 00530 matched production evidence

Run 00530 provides positive but not complete validation for the physical-grid temporal mapping introduced here. Relative to matched 00529:

- source-low upper45 first boundary vertical impulse, target-grid equivalent: **+0.239003 -> +0.190273 latent cells** (**-20.4%**);
- source-low upper45 four-step net vertical displacement, target-grid equivalent: **+0.298533 -> +0.216870** (**-27.4%**);
- learned-transfer upper45 first boundary dy: **+0.376029 -> +0.275541** (**-26.7%**);
- final post-high upper45 first boundary dy: **+0.484107 -> +0.282826** (**-41.6%**);
- final post-high upper45 four-step net dy: **+0.552679 -> +0.273206** (**-50.6%**);
- decoded PT212 upper45 first-three post-boundary median dy: approximately **+3.200 -> +2.466 px/frame** (**-22.9%**).

The exact-context substitution remains a numerical no-op at source grid, so the improvement is attributable upstream of learned transfer rather than to prefix replacement. The user also reports the visible frame shift is now subtler and that the fish turn at the beginning is qualitatively better.

This is enough evidence to **keep** the H3 physical-grid mapping candidate; it should no longer be treated as a neutral/failed hypothesis. It is not a complete root-cause closure because a residual source-low boundary impulse remains: upper45 pre-boundary median dy is about **-0.02033 target-equivalent cells/frame**, while the first post-boundary step is still **+0.19027**.

No further arithmetic change is folded into this PR based on 00530 alone.

### CODE REVIEW

**Comment by xmarre** [comment]
Coordinated Flow head advanced to `537901e945d4ff6596d00336d91e42685125267a` solely to add non-mutating transfer-splice diagnostics. VDN #19 remains unchanged at `1bc9f9cd0f685a28f84664b50951df8241501249`; no current seam evidence justifies a VDN source change. Flow exact-head CI is green.

**Comment by xmarre** [comment]
Coordinated Flow #49 head is now `179932a3ed616fe32ce45b75406464edb7e8ed04`, adding diagnostics-only multi-frame latent trajectory receipts. Exact-head Flow CI is green. VDN #19 remains unchanged; no VDN behavior change is justified before the matched trajectory rerun.

**Comment by xmarre** [comment]
Closed as superseded by the released exact-prefix VDN production path. The branch remains preserved because active arithmetic/Keyless research PRs may still use it as a technical base, but this PR is no longer an open production candidate.
